### PR TITLE
Use a builder pattern to configure VMM

### PR DESCRIPTION
**Pull request by AlexandreBrg**
This issue wasn't scheduled at first, but I was intrigued by the builder pattern, so I tried to implement it on lumper. 

I'm also working on the VirtIO implementation 🤓

### CODE REVIEW

**Review by sameo — COMMENTED** [review]
Nice, thanks @AlexandreBrg 

A few comments only.

Also, please use a commit message with the following format: `<subsystem>: <Title>` and follow the current convention. We should document that somehow (hint...)
For example: `vmm: Use builder pattern to configure the VMM`

File: src/main.rs [hunk 1]
```diff
@@ -24,7 +25,11 @@ struct VMMOpts {
 
     /// Stdout console file path
     #[clap(long)]
-    console: Option<String>
+    console: Option<String>,
+
+    /// Define a TAP interface name used to give network to the guest
+    #[clap(short, long)]
+    tap: Option<String>,
```
[the remark this answers is not in the document]
Comment (AlexandreBrg) [reply src/main.rs:32]: 👍

**Comment by AlexandreBrg** [comment]
Thanks for the review, I guess some people too will see your comment about adding a commit convention 😉

File: .gitignore [hunk 2]
```diff
@@ -8,3 +8,6 @@ Cargo.lock
 
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+# IntelliJ idea files
+.idea
```
Comment (sameo) [inline .gitignore:13]: This is an unrelated change. Either drop it or add a new commit for that.

File: src/main.rs [hunk 3]
```diff
@@ -36,19 +36,28 @@ pub enum Error {
...
+        // * Memory size (in MB)
+        // * Path to a Linux kernel
+        // * Optional path to console file
+        vmm.configure(opts.cpus, opts.memory, &opts.kernel, opts.console)
+            .map_err(Error::VmmConfigure)?;
+
+        Ok(vmm)
+    }
+}
+
```
Comment (sameo) [inline src/main.rs:53]: Why is this change part of this commit? The commit subject line says: `Use thiserror crate to format error properly` and you add a `VMM` constructor to it. This looks unrelated as well.

File: src/vmm/src/lib.rs [hunk 4]
```diff
@@ -102,6 +118,7 @@ impl VMM {
             serial: Arc::new(Mutex::new(
                 LumperSerial::new(Box::new(stdout())).map_err(Error::SerialCreation)?,
             )),
+            io_mgr: Arc::new(Mutex::new(IoManager::new())),
```
Comment (sameo) [inline src/vmm/src/lib.rs]: Seems unrelated as well.

File: src/vmm/src/devices/mod.rs [hunk 5]
```diff
@@ -1,3 +1,4 @@
 // SPDX-License-Identifier: Apache-2.0
 
 pub(crate) mod serial;
+pub mod net;
```
Comment (sameo) [inline src/vmm/src/devices/mod.rs]: Unrelated to the commit.

File: .gitignore (hunk 2 above)
Comment (AlexandreBrg) [reply .gitignore:13]: Ok!

File: src/vmm/src/devices/mod.rs (hunk 5 above)
Comment (AlexandreBrg) [reply src/vmm/src/devices/mod.rs]: Removed.

File: src/main.rs (hunk 3 above)
Comment (AlexandreBrg) [reply src/main.rs:53]: You are right, I changed the commit related to this change, it is now part of builder commit. I also removed `.tap(opts.tap)` call, as it is related to NetConfig commit.

File: src/vmm/src/lib.rs (hunk 4 above)
Comment (AlexandreBrg) [reply src/vmm/src/lib.rs]: Deleted it, it shouldn't have been committed

**Comment by AlexandreBrg** [comment]
You had me messing with Git a lot, I've never been doing this kind of thing (splitting a commit in two, moving code from one commit to another), really interesting exercise, ty.

File: src/vmm/src/lib.rs [hunk 6]
```diff
@@ -179,10 +182,7 @@ impl VMM {
         Ok(())
     }
 
-    pub fn configure_console(
-        &mut self,
-        console_path: Option<String>
-    ) -> Result<()> {
+    pub fn configure_console(&mut self, console_path: Option<String>) -> Result<()> {
```
Comment (sameo) [inline src/vmm/src/lib.rs:191]: Nit: This is a formatting change that should be avoided.

**Review by sameo — COMMENTED** [review]
It does look a lot better @AlexandreBrg, thanks.
I have one final nit, and we're good to go.

File: src/vmm/src/lib.rs (hunk 6 above)
Comment (AlexandreBrg) [reply src/vmm/src/lib.rs]: Should it be avoided because the commit isn't related to it, or because it should stay multi-line ?

I removed the change,

Comment (sameo) [reply src/vmm/src/lib.rs:191]: > Should it be avoided because the commit isn't related to it, or because it should stay multi-line ?

Because it's unrelated, and because cargo fmt would not let you do that I think.

> I removed the change.

Thanks!

> EDIT: As the GitHub workflow didn't pass because of this line, I created a new commit for the format fix.

**Review by sameo — APPROVED** [review]
